Detectors/EMCAL/calib/include/EMCALCalib/CellRecalibrator.h

```diff
@@ -0,0 +1,205 @@
+    CellTypeException() = default;
+
+    /// \brief Destructor
+    ~CellTypeException() noexcept final = default;
+
+    /// \brief Get error message of the exception
+    /// \return Error message
+    const char* what() const noexcept final
+    {
+      return "Only possible to calibrate cells of type high gain or low gain";
+    }
+  };
+
+  /// \brief Constructor
+  CellRecalibrator() = default;
+
+  /// \brief Destructor
+  ~CellRecalibrator() = default;
+
+  /// \brief Set the bad channel map
+  /// \param bcm Bad channel map to be applied
+  void setBadChannelMap(const BadChannelMap* bcm) { mBadChannelMap = bcm; }
+
+  /// \brief Set the time calibration params
+  /// \param tcp Time calibration params to be applied
+  void setTimeCalibration(const TimeCalibrationParams* tcp) { mTimeCalibration = tcp; }
+
+  /// \brief Set the gain calibration params
+  /// \param gcf Gain calibration factors to be applied
+  void setGainCalibration(const GainCalibrationFactors* gcf) { mGainCalibration = gcf; }
+
+  /// \brief Check if the bad channel calibration is enabled
+  /// \return True if the bad channel calibration is active (object available), false otherwise
+  bool hasBadChannelMap() const { return mBadChannelMap != nullptr; }
+
+  /// \brief Check if the time calibration is enabled
+  /// \return True if the time calibration is active (object available), false otherwise
+  bool hasTimeCalibration() const { return mTimeCalibration != nullptr; }
+
+  /// \brief Check if the energy calibration is enabled
+  /// \return True if the energy calibration is active (object available), false otherwise
+  bool hasGainCalibration() const { return mGainCalibration != nullptr; }
+
+  /// \brief Get bad channel map currently used in the calibrator
+  /// \return Current bad channel map (nullptr if not set)
+  const BadChannelMap* getBadChannelMap() const { return mBadChannelMap; }
+
+  /// \brief Get time calibration parameters currently used in the calibrator
+  /// \return Current time calibration parameters (nullptr if not set)
+  const TimeCalibrationParams* getTimeCalibration() const { return mTimeCalibration; }
+
+  /// \brief Get gain calibration factors currently used in the calibrator
+  /// \return Current gain calibration factors (nullptr if not set)
+  const GainCalibrationFactors* getGainCalibration() const { return mGainCalibration; }
+
+  /// \brief Calibrate single cell
+  ///
+  /// Applying all calibrations provided based on presence of calibration objects. Only
+  /// cells of type high-gain or low-gain can be calibrated. Since the cell can be rejected
+  /// by the bad channel calibration the return type is std::optional, where an empty optional
+  /// reflects rejected cells.
+  ///
+  /// \param inputcell Cell to calibrate
+  /// \return Calibrated cell (empty optional if rejected by the bad channel map)
+  /// \throw CellTypeException in case the inputcell is neither of type high gain nor low gain
+  template <typename T>
```

Comment:
Why using template and not std::optional<Cell>? It anyway works only for Cells

Reply:
The reason for templating the function is that we are planning to modify Florian's pull request to introduce a second uncompressed cell type as transient object with the same interface and the class should support both (the compressed one during the online stage for ED and QC, the transient one starting from the asynchonous stage). Obviously we found a necessity for this step.